### PR TITLE
Merge `RBE` changes 

### PR DESCRIPTION
**Pull request by mitogh**
Merge RBE recurrence changes back into `master`.

Ticket: https://central.tri.be/issues/129517

Related: https://github.com/moderntribe/the-events-calendar/pull/2531

### CODE REVIEW

**Comment by lucatume** [comment]
@bordoni should this target `release/B9.07`?

**Comment by lucatume** [comment]
@mitogh fix the tests and conflicts, please.

**Comment by bordoni** [comment]
@mitogh a couple things were missing here.
* [] Ticket Reference
* [] Changelog
* [] Tests passing

When is this needed for? We should try get it for start B19.08 dev cycle if we can.

File: src/Tribe/Repository.php [hunk 1]
```diff
@@ -1361,10 +1361,6 @@ public function get_ids() {
 			 */
 			return array();
 		}
-
-		$query->set( 'fields', 'ids' );
-
```
Comment (lucatume) [inline src/Tribe/Repository.php]: @mitogh why is this removed? We're in the `get_ids` method so it seems getting IDs only is the right thing to do.

File: src/Tribe/Repository.php [hunk 2]
```diff
@@ -1361,10 +1361,6 @@ public function get_ids() {
...
-		return $query->get_posts();
```
Comment (lucatume) [inline src/Tribe/Repository.php]: And why is this removed?

File: src/Tribe/Repository.php (hunk 1 above)
Comment (mitogh) [reply src/Tribe/Repository.php]: My bad just rolled back thanks.

File: src/Tribe/Repository.php (hunk 2 above)
Comment (mitogh) [reply src/Tribe/Repository.php]: Same as above 🙈

**Comment by mitogh** [comment]
Change log references to tickets will be updated once the main ticket is created.

**Comment by mitogh** [comment]
Would you mind to take a look at the PR: 

@sc0ttkclark || @lucatume || @bordoni 

Thank you!

File: src/Tribe/Repository.php [hunk 3]
```diff
@@ -1344,27 +1344,32 @@ public function get_ids() {
 			$query = $this->get_query();
 
 			// The request property will be set during the `get_posts` method and empty before it.
-			if ( ! empty( $query->request ) ) {
-				return array_map( static function ( $post ) {
+			if ( empty( $query->request ) ) {
+				$query->set( 'fields', 'ids' );
+
+				// Tests: 941, Assertions: 1592, Errors: 2, Failures: 44.
```
Comment (sc0ttkclark) [inline src/Tribe/Repository.php]: ?

File: src/Tribe/Utils/Array.php [hunk 4]
```diff
@@ -1,9 +1,13 @@
 <?php
+
+if ( class_exists( 'Tribe__Utils__Array' ) ) {
```
Comment (sc0ttkclark) [inline src/Tribe/Utils/Array.php]: I seem to remember you have to wrap the whole class definition in order to use this kind of logic. Is that still true? Have you tested overriding the class with this way? Maybe I'm remembering the PHP 5.x days, was it fixed in 7.x+

File: src/Tribe/Repository.php (hunk 3 above)
Comment (mitogh) [reply src/Tribe/Repository.php]: My bad 🙈

File: src/functions/template-tags/general.php [hunk 5]
```diff
@@ -752,3 +752,105 @@ function tribe_context() {
...
+	 * @param int|WP_Post $post_id   The post ID or object
+	 * @param array       $use_frags An array detailing the post fields, columns of
+	 *                               the `posts` table, that should be used to build
+	 *                               the checksum; defaults to the post `ID` and `post_modified`
+	 *                               columns.
+	 *
+	 * @return null|string The checksum string or `null` if the post is not valid or no
+	 *                     fragments are passed.
+	 */
+	function tribe_post_checksum( $post_id, $use_frags = array( 'ID', 'post_modified' ) ) {
```
Comment (sc0ttkclark) [inline src/functions/template-tags/general.php]:
```suggestion
	function tribe_post_checksum( $post_id, $use_frags = [ 'ID', 'post_modified' ] ) {
```

Also not sure if this default should be set in the function or within the function and check for `null === $use_frags`.

File: src/functions/template-tags/general.php [hunk 6]
```diff
@@ -752,3 +752,105 @@ function tribe_context() {
...
+	 *                               the checksum for.
+	 * @param array       $use_frags An array detailing the post fields, columns of
+	 *                               the `posts` table, that should be used to build
+	 *                               the checksum; defaults to the post `ID` and `post_modified`
+	 *                               columns.
+	 *
+	 * @return null|string The checksum string or `null` if the posts are not valid or no
+	 *                     fragments are passed.
+	 */
+	function tribe_posts_checksum( $post_ids, $use_frags = array( 'ID', 'post_modified' ) ) {
```
Comment (sc0ttkclark) [inline src/functions/template-tags/general.php]:
```suggestion
	function tribe_posts_checksum( $post_ids, $use_frags = [ 'ID', 'post_modified' ] ) {
```

Also not sure if this default should be set in the function or within the function and check for null === $use_frags.

File: src/functions/template-tags/general.php [hunk 7]
```diff
@@ -752,3 +752,105 @@ function tribe_context() {
...
+			/** @var wpdb $wpdb */
+			global $wpdb;
+			$prepared = array();
+			foreach ( $post_ids as $post_id ) {
+				if ( $post_id instanceof WP_Post ) {
+					$post_id = $post_id->ID;
+				}
+				$prepared[] = $wpdb->prepare( '%d', $post_id );
+			}
+			$fields = implode( ', ', array_map( 'esc_sql', $use_frags ) );
```
Comment (sc0ttkclark) [inline src/functions/template-tags/general.php]:
```suggestion
			$fields = '`' . implode( '`, `', array_map( 'esc_sql', $use_frags ) ) . '`';
```

File: src/Tribe/Utils/Array.php (hunk 4 above)
Comment (mitogh) [reply src/Tribe/Utils/Array.php]: Adding a test for it.

**Comment by mitogh** [comment]
Hey @sc0ttkclark would you mind to take another look at this one once you have a chance.

Thank you!

File: src/functions/template-tags/general.php [hunk 8]
```diff
@@ -807,7 +808,7 @@ function tribe_post_checksum( $post_id, $use_frags = array( 'ID', 'post_modified
 	 * @return null|string The checksum string or `null` if the posts are not valid or no
 	 *                     fragments are passed.
 	 */
-	function tribe_posts_checksum( $post_ids, $use_frags = array( 'ID', 'post_modified' ) ) {
+	function tribe_posts_checksum( $post_ids, $use_frags = [ 'ID', 'post_modified' ] ) {
```
Comment (sc0ttkclark) [inline src/functions/template-tags/general.php]: Should this also have the `null` usage as in `tribe_post_checksum()`?

**Comment by mitogh** [comment]
Keeping the branch so Loxi can point to `master` instead of this branch once we release a new hotfix.

**Comment by bordoni** [comment]
I have a feeling some tests in TEC might have broken.
https://travis-ci.org/moderntribe/the-events-calendar/builds/548306957#L1992

@mitogh would you mind checking that one?
